sync-core/src/main/java/com/cloudant/sync/query/QuerySqlTranslator.java

```diff
@@ -0,0 +1,386 @@
+ *  performing several queries over indexes and then using set operations works out
+ *  more flexible and likely more efficient.
+ *
+ *  The SQL must be executed separately so we can do it in a transaction so we're doing
+ *  it over a consistent view of the index.
+ *
+ *  The translator is a simple depth-first, recursive decent parser over the selector
+ *  map(dictionary).
+ *
+ *  Some examples:
+ *
+ *  AND : [ { x: X }, { y: Y } ]
+ *
+ *  This can be represented by a single SQL query and AND tree node:
+ *
+ *  AND
+ *   |
+ *  sql
+ *
+ *
+ *  OR : [ { x: X }, { y: Y } ]
+ *
+ *  This is a single OR node and two SQL queries:
+ *
+ *     OR
+ *     /\
+ *  sql sql
+ *
+ *  The interpreter then unions the results.
+ *
+ *
+ *  OR : [ { AND : [ { x: X }, { y: Y } ] }, { y: Y } ]
+ *
+ *  This requires a more complex tree:
+ *
+ *      OR
+ *     / \
+ *  AND  sql
+ *   |
+ *  sql
+ *
+ *  We could collapse out the extra AND node.
+ *
+ *
+ *  AND : [ { OR : [ { x: X }, { y: Y } ] }, { y: Y } ]
+ *
+ *  This is really the most complex situation:
+ *
+ *       AND
+ *       / \
+ *     OR  sql
+ *     /\
+ *  sql sql
+ *
+ *  These basic patterns can be composed into more complicate structures.
+ */
+class QuerySqlTranslator {
+
+    private static final String AND = "$and";
+    private static final String OR = "$or";
+    private static final String EQ = "$eq";
+    private static final String EXISTS = "$exists";
+
+    private static final Logger logger = Logger.getLogger(QuerySqlTranslator.class.getName());
+
+    public static QueryNode translateQuery(Map<String, Object> query,
+                                           Map<String, Object> indexes,
+                                           Boolean[] indexesCoverQuery) {
+        TranslatorState state = new TranslatorState();
+        QueryNode node = translateQuery(query, indexes, state);
+
+        // If we haven't used a single index, we need to return a query
+        // which returns every document, so the posthoc matcher can
+        // run over every document to manually carry out the query.
+        if (!state.atLeastOneIndexUsed) {
+            // TODO
+            return null;
+        } else {
+            indexesCoverQuery[0] = !state.atLeastOneIndexMissing;
+            return node;
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private static QueryNode translateQuery(Map<String, Object> query,
+                                           Map<String, Object> indexes,
+                                           TranslatorState state) {
+        // At this point we will have a root compound predicate, AND or OR, and
+        // the query will be reduced to a single entry:
+        // { "$and": [ ... predicates (possibly compound) ... ] }
+        // { "$or": [ ... predicates (possibly compound) ... ] }
+
+        ChildrenQueryNode root = null;
+        List<Object> clauses = new ArrayList<Object>();
+
+        if (query.keySet().toArray()[0].equals(AND)) {
```

Comment:
Why not use the fact that `Map.get(blah)` returns `null` for key not present, as the [objc code](https://github.com/cloudant/CloudantQueryObjc/blob/5432ddc8b68dc1ca7fc45225d84529df3614066f/Pod/Classes/CDTQQuerySqlTranslator.m#L123) does and you have elsewhere?